adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java

```diff
@@ -1226,4 +1300,110 @@ private static MqttConnectReturnCode getConnectReturnCode(final Throwable e) {
             return MqttConnectReturnCode.CONNECTION_REFUSED_NOT_AUTHORIZED;
         }
     }
+
+
+    /**
+     * A class that contains handlers for command subscription, unsubscription and sending of commands.
+     */
+    protected class CommandHandler {
+
+        private final Map<String, TriTuple<CommandSubscription, MessageConsumer, Object>> subscriptions = new ConcurrentHashMap<>();
+        private final Map<Integer, TriTuple<Long, CommandSubscription, CommandContext>> waitingForAcknowledgement = new ConcurrentHashMap<>();
+
+        /**
+         * Invoked when a device sends an MQTT <em>PUBACK</em> packet.
+         *
+         * @param msgId The msgId of the command published with QoS 1.
+         */
+        protected void handlePubAck(final Integer msgId) {
+            LOG.trace("Acknowledgement received for command [Msg-id: {}] that has been sent to device.", msgId);
+            Optional.ofNullable(removeFromWaitingForAcknowledgement(msgId)).ifPresent(value -> {
+                cancelTimer(value.one());
+
+                final CommandSubscription subscription = value.two();
```

Comment:
what if the timer has already fired? Do you still need to do all of the following in that case?

Reply:
- If the timer has already fired then there will be no entry for the msgId in map and hence ``ifPresent(value ->...)`` won't be invoked.

- After entering the ``ifPresent(value ->...)`` block and if the timer expires, the timer block finds no  entry for the msgId in map. I don't see any issue.

Reply:
Ah, I see